src/main/java/org/opensearch/knn/training/TrainingJobRunner.java

```diff
@@ -0,0 +1,174 @@
+    private static ModelDao modelDao;
+    private static ThreadPool threadPool;
+
+    private final Semaphore semaphore;
+    private final AtomicInteger jobCount;
+
+    /**
+     * Get singleton instance of TrainingJobRunner
+     *
+     * @return singleton instance of TrainingJobRunner
+     */
+    public static synchronized TrainingJobRunner getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new TrainingJobRunner();
+        }
+        return INSTANCE;
+    }
+
+    private TrainingJobRunner() {
+        this.jobCount = new AtomicInteger(0);
+        this.semaphore = new Semaphore(1);
+    }
+
+    /**
+     * Initializes static components.
+     *
+     * @param threadPool threadPool to use to get KNN Training Executor
+     * @param modelDao modelDao used to serialize the models
+     */
+    public static void initialize(ThreadPool threadPool, ModelDao modelDao) {
+        TrainingJobRunner.threadPool = threadPool;
+        TrainingJobRunner.modelDao = modelDao;
+    }
+
+    /**
+     * Execute a training job. This function will first grab a permit, and then serialize the initial model, then
+     * execute training, and then serialize the final result.
+     *
+     * @param trainingJob training job to be executed
+     * @param listener listener to handle final model serialization response (or exception)
+     */
+    public void execute(TrainingJob trainingJob, ActionListener<IndexResponse> listener) throws IOException {
+        // If the semaphore cannot be acquired, the node is unable to execute this job. This allows us to limit
+        // the number of training jobs that enter this function. Although the training threadpool size will also prevent
+        // this, we want to prevent this before we perform any serialization.
+        if (!semaphore.tryAcquire()) {
+            throw new RejectedExecutionException("Unable to run training job: No training capacity on node.");
+        }
+
+        jobCount.incrementAndGet();
+
+        // Serialize model before training. The model should be in the training state and the model binary should be
+        // null. This notifies users that their model is training, but not yet ready for use.
+        try {
+            serializeModel(
+                    trainingJob,
+                    ActionListener.wrap(
+                            indexResponse -> {
+                                // Update model id with id from response. We do this because users may or may not
+                                // provide an id
+                                trainingJob.setModelId(indexResponse.getId());
+                                train(trainingJob, listener);
+                            },
+                            exception -> {
+                                // Serialization failed. Let listener handle the exception, but free up resources.
+                                jobCount.decrementAndGet();
+                                semaphore.release();
```

Comment:
Looks like we are catching exception below and already doing these statements in line 113 and 114. Do we still need this?

Reply:
Those are outside of the handler. That occurs when the call to serializeModel fails before it can even submit the request. Here, we are handling the case when serializeModel submits the request, but the request fails.